api/profile/src/lib.rs

```diff
@@ -0,0 +1,88 @@
+mod model;
+mod config;
+mod utils;
+
+use anyhow::Result;
+use bytes::Bytes;
+use spin_sdk::{
+    http::{Request, Response},
+    http_component,
+};
+use config::Config;
+use utils::{bad_request,internal_server_error,method_not_allowed,not_found, ok, no_content};
+use model::Profile;
+
+enum Api {
+    Create(model::Profile),
+    ReadByHandle(String),
+    Update(model::Profile),
+    Delete(model::Profile),
+    BadRequest,
+    NotFound,
+    MethodNotAllowed,
+}
+
+#[http_component]
+fn profile_api(req: Request) -> Result<Response> {
+    let cfg = Config::get();
+
+    match api_from_request(req) {
+        Api::BadRequest => bad_request(),
+        Api::MethodNotAllowed => method_not_allowed(),
+        Api::Create(model) => handle_create(&cfg.db_url, model),
+        Api::Update(model) => handle_update(&cfg.db_url, model),
+        Api::ReadByHandle(handle) => handle_read_by_handle(&cfg.db_url, handle),
+        Api::Delete(handle) => handle_delete_by_handle(&cfg.db_url, handle),
+        _ => not_found(),
+    }
+}
+
+fn api_from_request(req: Request) -> Api {
+    match *req.method() {
+        http::Method::POST => match Profile::from_bytes(req.body().as_ref().unwrap_or(&Bytes::new())) {
+            Ok(model) => Api::Create(model),
+            Err(_) => Api::BadRequest,
+        }
+        http::Method::GET => match Profile::from_path(&req.headers()) {
+            Ok(model) => Api::ReadByHandle(model.handle),
+            Err(_) => Api::NotFound,
+        },
+        http::Method::PUT => match Profile::from_bytes(req.body().as_ref().unwrap_or(&Bytes::new())) {
+            Ok(model) => Api::Update(model),
+            Err(_) => Api::BadRequest,
+        },
+        http::Method::DELETE => match Profile::from_path(&req.headers()) {
+            Ok(model) => Api::Delete(model),
+            Err(_) => Api::NotFound,
+        },
+        _ => Api::MethodNotAllowed,
+    }
+}
```

Comment:
Nested matches are always a bit challenging to follow.

Maybe consider something like:
```rust
```rust
fn api_from_request(req: Request) -> Api {
    let profile = Profile::from_bytes(req.body().as_ref().unwrap_or(&Bytes::new()));

    match (*req.method(), profile) {
        (http::Method::Post, Ok(model)) => Api::Create(model),
        (http::Method::Post, Err(_e)) => {
            // log::debug!("{e}") here might be useful
            Api::BadRequest
        },
        // similarily for remaining cases...
    }
}

```

Reply:
Aha! I tried to flatten the nesting of match but there is something happening with borrowing in a match statement that I don't quite understand yet. Thanks!